### PR TITLE
fix device selection BUG

### PR DESCRIPTION
**Pull request by psychocoderHPC**
On multi gpu cluster nodes without `default` compute mode the device is normally selected by the gpu driver with the first cuda call.
It might be that this is not the case if the first call is `cudaStreamCreate`. 
In this case streams are created on the wrong device. This breaks our event system and triggers random crashes.

*This error was seen with CUDA 7.0rc, but it might be that it was not checked by the driver before.*
 
**Solution:**
- set device explicitly independent of the compute mode

**Tests**
- [x] runtime tests

### CODE REVIEW

**Comment by ax3l** [comment]
huh, interesting.

can we at the same time select devices that are "free" on the node? e.g. by testing them first.

imagine a node is not used exclusively and 2 of 4 devices are already used. I might start a specific jobscript looking for "2 gpus per node" or similar to utilize them.

**Comment by psychocoderHPC** [comment]
If `process` or `thread exclusive` mode is enabled we use only free devices. If the compute mode is set to default we can't check if a device is already used. It might be possible with NVML.

File: src/libPMacc/include/Environment.hpp [hunk 1]
```diff
@@ -38,160 +38,174 @@
...
 
-    template <unsigned DIM = DIM1>
-    class Environment
+    PMacc::GridController<DIM>& GridController()
     {
-    public:
+        return PMacc::GridController<DIM>::getInstance();
+    }
 
-        PMacc::GridController<DIM>& GridController()
```
Comment (ax3l) [inline src/libPMacc/include/Environment.hpp]: but whyyyyy? :it:

File: src/libPMacc/include/Environment.hpp [hunk 2]
```diff
@@ -38,160 +38,174 @@
...
+        else if (num_gpus < deviceNumber) //check if device can be selected by deviceNumber
         {
+            std::cerr << "no CUDA device " << deviceNumber << ", only " << num_gpus << " devices found" << std::endl;
+            throw std::runtime_error("CUDA capable devices can't be selected");
         }
 
-        Environment(const Environment&);
 
-        Environment& operator=(const Environment&);
+        int maxTrys = num_gpus;
```
Comment (ax3l) [inline src/libPMacc/include/Environment.hpp]: one try, [two tries](http://dict.tu-chemnitz.de/dings.cgi?lang=en&query=tries).

`maxTries = ...`

File: src/libPMacc/include/Environment.hpp [hunk 3]
```diff
@@ -38,160 +38,174 @@
...
 
-        void setDevice(int deviceNumber)
+        cudaDeviceProp devProp;
+        cudaError rc;
+        CUDA_CHECK(cudaGetDeviceProperties(&devProp, deviceNumber));
+        if (devProp.computeMode == cudaComputeModeDefault)
```
Comment (ax3l) [inline src/libPMacc/include/Environment.hpp]: minor style:
```c++
/*if the gpu compute mode is set to default we use the given `deviceNumber`*/
if (devProp.computeMode == cudaComputeModeDefault)
    maxTrys = 1;
```

File: src/libPMacc/include/Environment.hpp [hunk 4]
```diff
@@ -38,160 +38,174 @@
...
+            maxTrys = 1;
+        }
+
+        for (int deviceOffset = 0; deviceOffset < maxTrys; ++deviceOffset)
+        {
+            const int tryDeviceId = (deviceOffset + deviceNumber) % num_gpus;
+            rc = cudaSetDevice(tryDeviceId);
+            if (rc == cudaSuccess)
             {
-                std::cerr << "no CUDA device " << deviceNumber << ", only " << num_gpus << " devices found" << std::endl;
```
Comment (ax3l) [inline src/libPMacc/include/Environment.hpp]: I don't even understand why a `cudaSuccess` resulted in a `std::runtime_error` in the *old* code ... o.0

Comment (ax3l) [inline src/libPMacc/include/Environment.hpp]: oh, diff tricked me. all right.

**Comment by ax3l** [comment]
pew, the white space changes really force me to read that with `?w=1`...

**Comment by psychocoderHPC** [comment]
Yes sry for the white space changes. It time to use uncrustify :-)

**Comment by ax3l** [comment]
pls update the last inline comment so I can merge that :)

**Comment by ax3l** [comment]
pls merge the updates: https://github.com/psychocoderHPC/picongpu/pull/4

**Comment by psychocoderHPC** [comment]
thx merged

**Pull request merged**